cdk/backend.ts

```diff
@@ -163,23 +168,46 @@ const coapDockerfilePath = path.join(
 	'containers',
 	'coap',
 )
-const coapHash = await hashFolder(coapDockerfilePath)
+const coapSimulatorDownloadUrl = (
+	await getCoAPHealthCheckSettings({
+		ssm,
+		stackName: STACK_NAME,
+	})
+).simulatorDownloadURL.toString()
+const { body, ok, status } = await fetch(coapSimulatorDownloadUrl, {
+	method: 'GET',
+})
+if (!ok) throw new Error(`Failed to download CoAP simulator: ${status}`)
+
+const tmpCoapSimulatorPath = path.join(tmpdir(), 'coap-simulator.zip')
+await promisify(pipeline)(
+	body as unknown as NodeJS.ReadableStream,
+	createWriteStream(tmpCoapSimulatorPath),
+)
+
+const coapSimulatorBinaryHash = await hashFile(tmpCoapSimulatorPath)
+const coapDockerfileHash = await hashFolder(coapDockerfilePath)
+const coapHash = hashStrings([coapDockerfileHash, coapSimulatorBinaryHash])
 const { imageTag: coapSimulatorImageTag } = await getOrBuildDockerImage({
 	ecr,
 	releaseImageTag: undefined,
 	debug: debug('CoAP simulator docker image'),
+	beforeBuild: async () => {
+		await copyFile(
+			tmpCoapSimulatorPath,
+			path.join(coapDockerfilePath, 'coap-simulator.zip'),
+		)
+	},
+	afterBuild: async () => {
+		await rm(path.join(coapDockerfilePath, 'coap-simulator.zip'))
+	},
```

Comment:
Why is this hooked into the logic of getOrBuildDockerImage(), when `getOrBuildDockerImage` does nothing with these functions.

I'd like to see the logic separated to avoid this kind of workflow nesting.

1. Since we know based on the `coapHash` whether the Docker image needs to be rebuild (we can check whether the current docker image hash is the same as `coapHash`.
2. If not, download simulator ZIP and copy to expected file location for build (`copyeFile(...)`)
3. Build the new image
4. We don't need to delete `coap-simulator.zip`, so the `afterBuild` step can be removed.

Reply:
So, we have to execute function to check the existing tag twice, one before calling `getOrBuildDockerImage` and another one inside `getOrBuildDockerImage`.